src/walks.jl

```diff
@@ -0,0 +1,159 @@
+"""
+    AbstractWalk
+
+Any walk for use with [`fmap`](@ref) should inherit from this type.
+A walk subtyping `AbstractWalk` must satisfy the walk function interface:
+```julia
+struct MyWalk <: AbstractWalk end
+
+function (::MyWalk)(recurse, x, ys...)
+  # implement this
+end
+```
+The walk function is called on a node `x` in a Functors tree.
+It may also be passed associated nodes `ys...` in other Functors trees.
+The walk function recurses further into `(x, ys...)` by calling
+`recurse` on the child nodes.
+The choice of which nodes to recurse and in what order is custom to the walk.
+"""
+abstract type AbstractWalk end
+
+"""
+    AnonymousWalk(walk_fn)
+
+Wrap a `walk_fn` so that `AnonymousWalk(walk_fn) isa AbstractWalk`.
+This type only exists for backwards compatability and should be directly used.
+Attempting to wrap an existing `AbstractWalk` is a no-op (i.e. it is not wrapped).
+"""
+struct AnonymousWalk{F} <: AbstractWalk
```

Comment:
```suggestion
struct AnonymousWalk{F<:AbstractWalk} <: AbstractWalk
```
Is there any reason not to constrain these type parameters, at least for now?

Reply:
The reason this type exists is to wrap `F` that isn't a `AbstractWalk` but still satisfies the `AbstractWalk` interface.